pip/pip-333.md

```diff
@@ -0,0 +1,387 @@
+                        Set<String> ciphers, Set<String> protocols, String trustCertsFilePath,
+                        String certificateFilePath, String keyFilePath, boolean allowInsecureConnection,
+                        boolean requireTrustedClientCertOnConnect, AuthenticationDataProvider authData,
+                        boolean isKeyStoreEnabled);
+  public SSLEngine getClientSslEngine(String peerHost, int peerPort);
+  public SSLEngine getServerSslEngine();
+  public boolean needsUpdate();
+  public void update() throws Exception;
+  //The below will be a noop
+  public void configure(String sslProviderString,
+                        Set<String> ciphers,
+                        Set<String> protocols,
+                        boolean allowInsecureConnection,
+                        boolean requireTrustedClientCertOnConnect,
+                        AuthenticationDataProvider authData,
+                        String tlsCustomParams);
+}
+```
+
+### Pulsar Commmon Changes
+
+6 new configurations will need to be added into the Configurations like `ServiceConfiguration`, 
+`ClientConfigurationData`, `ProxyConfiguration`, etc. All of the below will be optional. It will use the default values
+to match the current behavior of Pulsar.
+
+- `sslFactoryPlugin`: SSL Factory Plugin class to provide SSLEngine and SSLContext objects.
+The default class used is `DefaultSslFactory`.
+- `sslFactoryPluginParams`: SSL Factory plugin configuration parameters. It will be of type string. It can be parsed by
+the plugin at its discretion.
+- `sslFactorySslContextInitTimeoutInMs`: SSL Context initialization timeout in ms for the first time it is initialized 
+by the SSLFactory
+
+The below configs will be applicable only to the Pulsar Server components like Broker and Proxy:
+- `brokerClientSslFactoryPlugin`: SSL Factory Plugin class used by internal client to provide SSLEngine and SSLContext 
+objects. The default class used is `DefaultSslFactory`.
+- `brokerClientSslFactoryPluginParams`: SSL Factory plugin configuration parameters used by internal client. It can be 
+parsed by the plugin at its discretion.
+- `brokerClientsslFactorySslContextInitTimeoutInMs`: SSL Context initialization timeout in ms for the first time 
+it is initialized by the SSLFactory in the internal client
+
+`JettySslContextFactory` class will need to be changed to internally use the `SslFactory` class to generate the 
+SslContext.
+
+### SslFactory Usage across Pulsar Netty based server components
+
+Example Changes in broker's `PulsarChannelInitializer` to initialize the SslFactory:
+```java
+sslFactory = (SslFactory) Class.forName(config.getSslFactoryPlugin())
+                    .getDeclaredConstructor(Long.TYPE, LONG.TYPE)
+                    .newInstance(config.getTlsCertRefreshCheckDurationSec(),
+                                 config.getSslFactorySslContextInitTimeoutInMs);
+if (this.sslFactory instanceof DefaultSslFactory) {
```

Comment:
Please see the previous comment of having a solution where the PulsarSslFactory interface's confgiure method would pass in a value object with all possible configuration options. In that case there wouldn't need to be any conditional login here.

Reply:
Made the modifications.